src/arch/arm/kernel/boot.c

```diff
@@ -315,7 +315,27 @@ static BOOT_CODE bool_t try_init_kernel(
     create_frames_of_region_ret_t create_frames_ret;
     create_frames_of_region_ret_t extra_bi_ret;
 
-    /* convert from physical addresses to userland vptrs */
+    /* Convert from physical addresses to userland vptrs with the parameter
+     * pv_offset, which is defined as:
+     *     virt_address + pv_offset = phys_address
+     * The offset is usually a positive value, because the virtual address of
+     * the user image is a low value and the actually physical address is much
+     * greater. However, there is no restrictions on the physical and virtual
+     * image location in general. Defining pv_offset as a signed value might
+     * seem the intuitive choice how to handle this, but there is are two
+     * catches here that break the C rules. We can't cover the full integer
+     * range then, and overflows/underflows are well defined for unsigned values
+     * only. They are undefined for signed values, even if such operations
+     * practically work in many cases due to how compiler and machine implement
+     * negative integers using the two's-complement.
+     * Assume a 32-bit system with virt_address=0xc0000000 and phys_address=0,
```

Comment:
I do not quite get the example here.

Reply:
What exactly is unclear here? We assume we are running vom phys address 0 and are linked on virtual address 0xc000000. The calculations will then leave the signed integer range, that is why we can't use signed integers here. Technically, things still works because of the 2s-complement number representation. In the end all this is about staying in the valid C domain, because over/underflow on unsigned values is undefined (even if it likely works)